File: modules/file-engine/ios/ImageText.swift
import UIKit
import Vision
import ImageIO
import UniformTypeIdentifiers

/// On-device text recognition (Apple Vision) and redraw. Coordinates are normalised to the upright image.
enum ImageText {
  private static let analysisSize: CGFloat = 2048
  private static let maxLines = 400

  private struct Pixels {
    let data: [UInt8]; let width: Int; let height: Int
    init?(_ image: CGImage) {
      width = image.width; height = image.height
      var buffer = [UInt8](repeating: 0, count: width * height * 4)
      guard let context = CGContext(data: &buffer, width: width, height: height, bitsPerComponent: 8, bytesPerRow: width * 4,
                                    space: CGColorSpaceCreateDeviceRGB(), bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else { return nil }
      context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
      data = buffer
    }
    func at(_ x: Int, _ y: Int) -> Int {
      let i = (min(max(y, 0), height - 1) * width + min(max(x, 0), width - 1)) * 4
      return Int(data[i]) << 16 | Int(data[i + 1]) << 8 | Int(data[i + 2])
    }
  }

  static func recognize(uri: String) throws -> [String: Any] {
    guard let image = ImageEditing.load(uri: uri, maxPixels: analysisSize) else { throw ImageEditing.EditError.invalid("This image format cannot be read on your device.") }
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    try VNImageRequestHandler(cgImage: image, options: [:]).perform([request])
    let pixels = Pixels(image)
    var lines: [[String: Any]] = []
    for observation in request.results ?? [] {
      guard lines.count < maxLines, let candidate = observation.topCandidates(1).first, !candidate.string.trimmingCharacters(in: .whitespaces).isEmpty else { continue }
      let box = observation.boundingBox
      let rect = CGRect(x: box.minX, y: 1 - box.maxY, width: box.width, height: box.height)
      var line: [String: Any] = ["id": lines.count, "text": candidate.string, "x": rect.minX, "y": rect.minY, "width": rect.width, "height": rect.height, "angle": 0]
      if let pixels {
        let colors = sample(pixels, CGRect(x: rect.minX * CGFloat(pixels.width), y: rect.minY * CGFloat(pixels.height), width: rect.width * CGFloat(pixels.width), height: rect.height * CGFloat(pixels.height)))
        line["color"] = colors.text; line["background"] = colors.background
        line["backgroundLeft"] = colors.left; line["backgroundRight"] = colors.right
      } else {
        line["color"] = 0x101010; line["background"] = 0xFFFFFF; line["backgroundLeft"] = 0xFFFFFF; line["backgroundRight"] = 0xFFFFFF
      }
      lines.append(line)
    }
    return ["width": image.width, "height": image.height, "lines": lines]
  }

  private static func channel(_ color: Int, _ shift: Int) -> Int { color >> shift & 255 }
  private static func median(_ colors: [Int]) -> Int {
    guard !colors.isEmpty else { return 0xFFFFFF }
    func m(_ shift: Int) -> Int { colors.map { channel($0, shift) }.sorted()[colors.count / 2] }
    return m(16) << 16 | m(8) << 8 | m(0)
  }
  private static func average(_ colors: [Int]) -> Int {
    guard !colors.isEmpty else { return 0 }
    func a(_ shift: Int) -> Int { colors.reduce(0) { $0 + channel($1, shift) } / colors.count }
    return a(16) << 16 | a(8) << 8 | a(0)
  }
  private static func distance(_ a: Int, _ b: Int) -> Double {
    let dr = Double(channel(a, 16) - channel(b, 16)), dg = Double(channel(a, 8) - channel(b, 8)), db = Double(channel(a, 0) - channel(b, 0))
    return (dr * dr + dg * dg + db * db).squareRoot()
  }

  /// Median of the pixels around the box is the background; the pixels farthest from it are the ink.
  private static func sample(_ pixels: Pixels, _ rect: CGRect) -> (background: Int, left: Int, right: Int, text: Int) {
    let pad = max(2, Int(rect.height * 0.18))
    let l = Int(rect.minX) - pad, r = Int(rect.maxX) + pad, t = Int(rect.minY) - pad, b = Int(rect.maxY) + pad
    var border: [Int] = [], leftEdge: [Int] = [], rightEdge: [Int] = []
    for x in stride(from: l, through: r, by: max(1, (r - l) / 80)) { border.append(pixels.at(x, t)); border.append(pixels.at(x, b)) }
    for y in stride(from: t, through: b, by: max(1, (b - t) / 30)) { leftEdge.append(pixels.at(l, y)); rightEdge.append(pixels.at(r, y)) }
    border += leftEdge + rightEdge
    let background = median(border)
    var inner: [Int] = []
    for y in stride(from: Int(rect.minY), through: Int(rect.maxY), by: max(1, Int(rect.height) / 20)) {
      for x in stride(from: Int(rect.minX), through: Int(rect.maxX), by: max(1, Int(rect.width) / 60)) { inner.append(pixels.at(x, y)) }
    }
    let distances = inner.map { distance($0, background) }
    let farthest = distances.max() ?? 0
    let text: Int
    if farthest < 40 {
      let luminance = (0.299 * Double(channel(background, 16)) + 0.587 * Double(channel(background, 8)) + 0.114 * Double(channel(background, 0))) / 255
      text = luminance > 0.5 ? 0x101010 : 0xFFFFFF
    } else {
      text = average(zip(inner, distances).filter { $0.1 >= farthest * 0.6 }.map { $0.0 })
    }
    return (background, median(leftEdge.isEmpty ? border : leftEdge), median(rightEdge.isEmpty ? border : rightEdge), text)
  }

  /// Maps the standard PDF font names (Helvetica, Times, Courier with Bold/Oblique/Italic) to iOS faces.
  static func font(_ name: String, size: CGFloat) -> UIFont {
    let bold = name.contains("Bold"), italic = name.contains("Oblique") || name.contains("Italic")
    let face: String
    if name.hasPrefix("Times") {
      face = bold && italic ? "TimesNewRomanPS-BoldItalicMT" : bold ? "TimesNewRomanPS-BoldMT" : italic ? "TimesNewRomanPS-ItalicMT" : "TimesNewRomanPSMT"
    } else {
      let family = name.hasPrefix("Courier") ? "Courier" : "Helvetica"
      face = family + (bold && italic ? "-BoldOblique" : bold ? "-Bold" : italic ? "-Oblique" : "")
    }
    if let font = UIFont(name: face, size: size) { return font }
    var traits: UIFontDescriptor.SymbolicTraits = []
    if bold { traits.insert(.traitBold) }
    if italic { traits.insert(.traitItalic) }
    let system = UIFont.systemFont(ofSize: size)
    return system.fontDescriptor.withSymbolicTraits(traits).map { UIFont(descriptor: $0, size: size) } ?? system
  }
  private static func color(_ value: Int) -> UIColor {
    UIColor(red: CGFloat(value >> 16 & 255) / 255, green: CGFloat(value >> 8 & 255) / 255, blue: CGFloat(value & 255) / 255, alpha: 1)
  }

  /// Covers replaced lines with the sampled background and draws text at its baseline.
  /// Sizes are in pixels of an image `refWidth` wide, so previews and full exports match.
  static func render(_ options: [String: Any]) throws -> [String: Any] {
    guard let uri = options["uri"] as? String, let output = options["outputUri"] as? String, let destination = URL(string: output), destination.isFileURL else {
      throw ImageEditing.EditError.invalid("Choose an image to edit.")
    }
    let roots = [FileManager.SearchPathDirectory.documentDirectory, .cachesDirectory].map { FileManager.default.urls(for: $0, in: .userDomainMask)[0].resolvingSymlinksInPath().path }
    let path = destination.resolvingSymlinksInPath().path
    guard roots.contains(where: { path.hasPrefix($0 + "/") }), !FileManager.default.fileExists(atPath: destination.path) else {
      throw ImageEditing.EditError.invalid("Could not save inside Versara.")
    }
    let preview = options["preview"] as? Bool ?? false
    let lowMemory = ProcessInfo.processInfo.physicalMemory <= 2 * 1024 * 1024 * 1024
    let limit = CGFloat(min(max(options["maxSize"] as? Int ?? (lowMemory ? 3072 : 4096), 256), 4096))
    guard let source = ImageEditing.load(uri: uri, maxPixels: limit) else { throw ImageEditing.EditError.invalid("This image format cannot be edited on your device.") }
    let edits = options["edits"] as? [[String: Any]] ?? []
    guard edits.count <= 500 else { throw ImageEditing.EditError.invalid("Save these changes before adding more.") }
    let size = CGSize(width: source.width, height: source.height)
    let ratio = size.width / max(1, CGFloat((options["refWidth"] as? NSNumber)?.doubleValue ?? Double(size.width)))
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = true
    let rendered = UIGraphicsImageRenderer(size: size, format: format).image { context in
      UIImage(cgImage: source).draw(in: CGRect(origin: .zero, size: size))
      let cg = context.cgContext
      for edit in edits {
        if let box = edit["erase"] as? [String: Any] {
          let x = (box["x"] as? NSNumber)?.doubleValue ?? 0, y = (box["y"] as? NSNumber)?.doubleValue ?? 0
          let w = (box["width"] as? NSNumber)?.doubleValue ?? 0, h = (box["height"] as? NSNumber)?.doubleValue ?? 0
          var rect = CGRect(x: x * size.width, y: y * size.height, width: w * size.width, height: h * size.height)
          let pad = max(2, rect.height * 0.15)
          rect = rect.insetBy(dx: -pad, dy: -pad)
          let fallback = box["background"] as? Int ?? 0xFFFFFF
          let colors = [color(box["left"] as? Int ?? fallback).cgColor, color(box["right"] as? Int ?? fallback).cgColor] as CFArray
          if let gradient = CGGradient(colorsSpace: CGColorSpaceCreateDeviceRGB(), colors: colors, locations: [0, 1]) {
            cg.saveGState()
            cg.addPath(UIBezierPath(roundedRect: rect, cornerRadius: pad).cgPath)
            cg.clip()
            cg.drawLinearGradient(gradient, start: CGPoint(x: rect.minX, y: 0), end: CGPoint(x: rect.maxX, y: 0), options: [.drawsBeforeStartLocation, .drawsAfterEndLocation])
            cg.restoreGState()
          }
        }
        if let text = edit["text"] as? String, !text.trimmingCharacters(in: .whitespaces).isEmpty {
          let points = CGFloat((edit["size"] as? NSNumber)?.doubleValue ?? 16) * ratio
          let face = font(edit["font"] as? String ?? "Helvetica", size: max(1, points))
          let x = CGFloat((edit["x"] as? NSNumber)?.doubleValue ?? 0) * size.width
          let baseline = CGFloat((edit["y"] as? NSNumber)?.doubleValue ?? 0) * size.height
          var attributes: [NSAttributedString.Key: Any] = [.font: face, .foregroundColor: color(edit["color"] as? Int ?? 0x101010)]
          if edit["underline"] as? Bool == true { attributes[.underlineStyle] = NSUnderlineStyle.single.rawValue }
          for (line, value) in text.components(separatedBy: "\n").prefix(50).enumerated() where !value.isEmpty {
            (value as NSString).draw(at: CGPoint(x: x, y: baseline + CGFloat(line) * face.pointSize * 1.2 - face.ascender), withAttributes: attributes)
          }
        }
      }
    }
    guard let image = rendered.cgImage else { throw ImageEditing.EditError.invalid("Could not render the image.") }
    let type: UTType = !preview && (options["format"] as? String) == "png" ? .png : .jpeg
    let quality = preview ? 0.85 : Double(min(max(options["quality"] as? Int ?? 92, 10), 100)) / 100
    try FileManager.default.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
    guard let writer = CGImageDestinationCreateWithURL(destination as CFURL, type.identifier as CFString, 1, nil) else { throw ImageEditing.EditError.invalid("Could not create the image file.") }
    CGImageDestinationAddImage(writer, image, [kCGImageDestinationLossyCompressionQuality: quality] as CFDictionary)
    guard CGImageDestinationFinalize(writer) else {
      try? FileManager.default.removeItem(at: destination)
      throw ImageEditing.EditError.invalid("Could not encode the image.")
    }
    let bytes = (try? destination.resourceValues(forKeys: [.fileSizeKey]).fileSize).map(Int64.init) ?? 0
    return ["uri": output, "width": image.width, "height": image.height, "size": bytes, "mimeType": type == .png ? "image/png" : "image/jpeg"]
  }
}

File: modules/file-engine/android/src/main/java/expo/modules/fileengine/ImageText.kt
package expo.modules.fileengine

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.Typeface
import android.net.Uri
import android.os.Build
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import org.json.JSONObject
import java.io.File
import kotlin.math.max
import kotlin.math.roundToInt
import kotlin.math.sqrt

/** On-device text recognition and redraw for images. Coordinates are normalised to the upright image. */
internal object ImageText {
  private const val ANALYSIS_SIZE = 2048
  private const val MAX_LINES = 400

  private class Sample(val background: Int, val left: Int, val right: Int, val text: Int)

  fun recognize(context: Context, uri: String): Map<String, Any> {
    val bitmap = ImageProcessing.decode(context, Uri.parse(uri), ANALYSIS_SIZE)
    val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    try {
      val result = Tasks.await(recognizer.process(InputImage.fromBitmap(bitmap, 0)))
      val w = bitmap.width.toFloat(); val h = bitmap.height.toFloat()
      val lines = ArrayList<Map<String, Any>>()
      for (block in result.textBlocks) for (line in block.lines) {
        if (lines.size >= MAX_LINES) break
        val box = line.boundingBox ?: continue
        if (box.width() < 2 || box.height() < 2 || line.text.isBlank()) continue
        val sample = sample(bitmap, box.left, box.top, box.right, box.bottom)
        lines += mapOf(
          "id" to lines.size, "text" to line.text,
          "x" to box.left / w.toDouble(), "y" to box.top / h.toDouble(), "width" to box.width() / w.toDouble(), "height" to box.height() / h.toDouble(),
          "angle" to line.angle.toDouble(),
          "color" to (sample.text and 0xFFFFFF), "background" to (sample.background and 0xFFFFFF),
          "backgroundLeft" to (sample.left and 0xFFFFFF), "backgroundRight" to (sample.right and 0xFFFFFF),
        )
      }
      return mapOf("width" to bitmap.width, "height" to bitmap.height, "lines" to lines)
    } finally {
      recognizer.close()
      bitmap.recycle()
    }
  }

  /** Median of the pixels around the box is the background; the pixels farthest from it are the ink. */
  private fun sample(bitmap: Bitmap, left: Int, top: Int, right: Int, bottom: Int): Sample {
    val pad = max(2, ((bottom - top) * 0.18f).roundToInt())
    val l = (left - pad).coerceIn(0, bitmap.width - 1); val r = (right + pad).coerceIn(0, bitmap.width - 1)
    val t = (top - pad).coerceIn(0, bitmap.height - 1); val b = (bottom + pad).coerceIn(0, bitmap.height - 1)
    val border = ArrayList<Int>(); val leftEdge = ArrayList<Int>(); val rightEdge = ArrayList<Int>()
    val stepX = max(1, (r - l) / 80); val stepY = max(1, (b - t) / 30)
    for (x in l..r step stepX) { border += bitmap.getPixel(x, t); border += bitmap.getPixel(x, b) }
    for (y in t..b step stepY) { leftEdge += bitmap.getPixel(l, y); rightEdge += bitmap.getPixel(r, y) }
    border += leftEdge; border += rightEdge
    val background = median(border)
    val inner = ArrayList<Int>()
    val ix = max(1, (right - left) / 60); val iy = max(1, (bottom - top) / 20)
    for (y in top.coerceIn(0, bitmap.height - 1)..bottom.coerceIn(0, bitmap.height - 1) step iy)
      for (x in left.coerceIn(0, bitmap.width - 1)..right.coerceIn(0, bitmap.width - 1) step ix) inner += bitmap.getPixel(x, y)
    val distances = inner.map { distance(it, background) }
    val farthest = distances.maxOrNull() ?: 0f
    val text = if (farthest < 40f) {
      if (luminance(background) > 0.5f) 0xFF101010.toInt() else 0xFFFFFFFF.toInt()
    } else {
      val ink = inner.filterIndexed { index, _ -> distances[index] >= farthest * 0.6f }
      average(ink)
    }
    return Sample(background, median(leftEdge.ifEmpty { border }), median(rightEdge.ifEmpty { border }), text)
  }

  private fun channel(color: Int, shift: Int) = color shr shift and 255
  private fun median(colors: List<Int>): Int {
    if (colors.isEmpty()) return 0xFFFFFFFF.toInt()
    fun m(shift: Int) = colors.map { channel(it, shift) }.sorted()[colors.size / 2]
    return (0xFF shl 24) or (m(16) shl 16) or (m(8) shl 8) or m(0)
  }
  private fun average(colors: List<Int>): Int {
    if (colors.isEmpty()) return 0xFF000000.toInt()
    fun a(shift: Int) = colors.sumOf { channel(it, shift) } / colors.size
    return (0xFF shl 24) or (a(16) shl 16) or (a(8) shl 8) or a(0)
  }
  private fun distance(a: Int, b: Int): Float {
    val dr = channel(a, 16) - channel(b, 16); val dg = channel(a, 8) - channel(b, 8); val db = channel(a, 0) - channel(b, 0)
    return sqrt((dr * dr + dg * dg + db * db).toFloat())
  }
  private fun luminance(color: Int) = (0.299f * channel(color, 16) + 0.587f * channel(color, 8) + 0.114f * channel(color, 0)) / 255f

  /** Maps the standard PDF font names (Helvetica, Times, Courier with Bold/Oblique/Italic) to system faces. */
  fun typeface(font: String): Typeface {
    val base = when {
      font.startsWith("Times") -> Typeface.SERIF
      font.startsWith("Courier") -> Typeface.MONOSPACE
      else -> Typeface.SANS_SERIF
    }
    val bold = font.contains("Bold")
    val italic = font.contains("Oblique") || font.contains("Italic")
    return Typeface.create(base, when { bold && italic -> Typeface.BOLD_ITALIC; bold -> Typeface.BOLD; italic -> Typeface.ITALIC; else -> Typeface.NORMAL })
  }

  /**
   * Covers replaced lines with the sampled background and draws text at its baseline.
   * Sizes are in pixels of an image [refWidth] wide, so previews and full exports match.
   */
  fun render(context: Context, options: JSONObject): Map<String, Any> {
    val destination = Uri.parse(options.getString("outputUri"))
    require(destination.scheme == "file") { "Destination must be an app file URI." }
    val target = File(requireNotNull(destination.path)).canonicalFile
    val allowed = listOf(context.filesDir.canonicalPath, context.cacheDir.canonicalPath)
    require(allowed.any { target.path.startsWith("$it/") }) { "Destination must stay inside the app." }
    require(!target.exists()) { "Destination already exists." }
    val preview = options.optBoolean("preview", false)
    val limit = options.optInt("maxSize", if (ImageProcessing.lowMemory(context)) 3072 else 4096).coerceIn(256, 4096)
    val decoded = ImageProcessing.decode(context, Uri.parse(options.getString("uri")), limit)
    val bitmap = if (decoded.isMutable && decoded.config == Bitmap.Config.ARGB_8888) decoded
    else decoded.copy(Bitmap.Config.ARGB_8888, true).also { decoded.recycle() }
    try {
      val canvas = Canvas(bitmap)
      val w = bitmap.width.toFloat(); val h = bitmap.height.toFloat()
      val ratio = w / options.optDouble("refWidth", w.toDouble()).toFloat().coerceAtLeast(1f)
      val edits = options.optJSONArray("edits")
      require((edits?.length() ?: 0) <= 500) { "Save these changes before adding more." }
      val fill = Paint(Paint.ANTI_ALIAS_FLAG)
      val ink = Paint(Paint.ANTI_ALIAS_FLAG or Paint.SUBPIXEL_TEXT_FLAG)
      for (index in 0 until (edits?.length() ?: 0)) {
        val edit = edits!!.getJSONObject(index)
        edit.optJSONObject("erase")?.let { box ->
          val rect = RectF((box.optDouble("x") * w).toFloat(), (box.optDouble("y") * h).toFloat(), ((box.optDouble("x") + box.optDouble("width")) * w).toFloat(), ((box.optDouble("y") + box.optDouble("height")) * h).toFloat())
          val pad = max(2f, rect.height() * 0.15f)
          rect.inset(-pad, -pad)
          val left = 0xFF000000.toInt() or box.optInt("left", box.optInt("background", 0xFFFFFF))
          val right = 0xFF000000.toInt() or box.optInt("right", box.optInt("background", 0xFFFFFF))
          fill.shader = LinearGradient(rect.left, 0f, rect.right, 0f, left, right, Shader.TileMode.CLAMP)
          canvas.drawRoundRect(rect, pad, pad, fill)
        }
        val text = edit.optString("text")
        if (text.isNotBlank()) {
          ink.typeface = typeface(edit.optString("font", "Helvetica"))
          ink.textSize = max(1f, edit.optDouble("size", 16.0).toFloat() * ratio)
          ink.color = 0xFF000000.toInt() or edit.optInt("color", 0x101010)
          ink.isUnderlineText = edit.optBoolean("underline", false)
          val x = (edit.optDouble("x") * w).toFloat()
          val y = (edit.optDouble("y") * h).toFloat()
          text.split('\n').take(50).forEachIndexed { line, value ->
            if (value.isNotEmpty()) canvas.drawText(value, x, y + line * ink.textSize * 1.2f, ink)
          }
        }
      }
      val format = if (preview) "jpeg" else options.optString("format", "jpeg")
      val quality = if (preview) 85 else options.optInt("quality", 92).coerceIn(10, 100)
      val (compress, mime) = when (format) {
        "png" -> Bitmap.CompressFormat.PNG to "image/png"
        "webp" -> (if (Build.VERSION.SDK_INT >= 30) Bitmap.CompressFormat.WEBP_LOSSY else @Suppress("DEPRECATION") Bitmap.CompressFormat.WEBP) to "image/webp"
        else -> Bitmap.CompressFormat.JPEG to "image/jpeg"
      }
      target.parentFile?.mkdirs()
      try { target.outputStream().use { require(bitmap.compress(compress, quality, it)) { "Could not encode the image." } } }
      catch (error: Throwable) { target.delete(); throw error }
      return mapOf("uri" to destination.toString(), "width" to bitmap.width, "height" to bitmap.height, "size" to target.length(), "mimeType" to mime)
    } finally {
      bitmap.recycle()
    }
  }
}
